layer4/routes.go

```diff
@@ -92,75 +98,71 @@ func (routes RouteList) Provision(ctx caddy.Context) error {
 // Compile prepares a middleware chain from the route list.
 // This should only be done once: after all the routes have
 // been provisioned, and before the server loop begins.
-func (routes RouteList) Compile(next Handler, logger *zap.Logger) Handler {
-	mid := make([]Middleware, 0, len(routes))
-	for _, route := range routes {
-		mid = append(mid, wrapRoute(route, logger))
-	}
-	stack := next
-	for i := len(mid) - 1; i >= 0; i-- {
-		stack = mid[i](stack)
-	}
-	return stack
-}
-
-// wrapRoute wraps route with a middleware and handler so that it can
-// be chained in and defer evaluation of its matchers to request-time.
-// Like wrapMiddleware, it is vital that this wrapping takes place in
-// its own stack frame so as to not overwrite the reference to the
-// intended route by looping and changing the reference each time.
-func wrapRoute(route *Route, logger *zap.Logger) Middleware {
-	return func(next Handler) Handler {
-		return HandlerFunc(func(cx *Connection) error {
-			// TODO: Update this comment, it seems we've moved the copy into the handler?
-			// copy the next handler (it's an interface, so it's just
-			// a very lightweight copy of a pointer); this is important
-			// because this is a closure to the func below, which
-			// re-assigns the value as it compiles the middleware stack;
-			// if we don't make this copy, we'd affect the underlying
-			// pointer for all future request (yikes); we could
-			// alternatively solve this by moving the func below out of
-			// this closure and into a standalone package-level func,
-			// but I just thought this made more sense
-			nextCopy := next
-
-			// route must match at least one of the matcher sets
-			matched, err := route.matcherSets.AnyMatch(cx)
+func (routes RouteList) Compile(next NextHandler, logger *zap.Logger, matchingTimeout time.Duration) Handler {
```

Comment:
I wonder if the aesthetics of calling this would be improved a bit if `next` is the last argument. That way we don't have a potential multi-line, inlined handler method that is followed by 2 more arguments after it.